File: test/lua/modules/linear_boltzmann_solvers/transport_steady/angular_io_1d.lua
-- Standard Reed 1D 1-group problem
-- Create Mesh
widths = { 2., 1., 2., 1., 2. }
nrefs = { 200, 200, 200, 200, 200 }

Nmat = #widths

nodes = {}
counter = 1
nodes[counter] = 0.
for imat = 1, Nmat do
  dx = widths[imat] / nrefs[imat]
  for i = 1, nrefs[imat] do
    counter = counter + 1
    nodes[counter] = nodes[counter - 1] + dx
  end
end

meshgen = mesh.OrthogonalMeshGenerator.Create({ node_sets = { nodes } })
grid = meshgen:Execute(meshgen)

-- Set Material IDs
z_min = 0.0
z_max = widths[1]
for imat = 1, Nmat do
  z_max = z_min + widths[imat]
  log.Log(LOG_0, "imat=" .. imat .. ", zmin=" .. z_min .. ", zmax=" .. z_max)
  lv = logvol.RPPLogicalVolume.Create({ infx = true, infy = true, zmin = z_min, zmax = z_max })
  grid:SetBlockIDFromLogicalVolume(lv, imat - 1, true)
  z_min = z_max
end

-- Add cross sections to materials
total = { 50., 5., 0., 1., 1. }
c = { 0., 0., 0., 0.9, 0.9 }
xs_map = {}
for imat = 1, Nmat do
  xs_map[imat] = {
    block_ids = { imat - 1 },
    xs = xs.CreateSimpleOneGroup(total[imat], c[imat]),
  }
end

-- Create sources in 1st and 4th materials
src0 = lbs.VolumetricSource.Create({ block_ids = { 0 }, group_strength = { 50. } })
src1 = lbs.VolumetricSource.Create({ block_ids = { 3 }, group_strength = { 1. } })

-- Angular Quadrature
gl_quad = aquad.CreateGLProductQuadrature1DSlab(128)

-- LBS block option
num_groups = 1
lbs_block = {
  mesh = grid,
  num_groups = num_groups,
  groupsets = {
    {
      groups_from_to = { 0, num_groups - 1 },
      angular_quadrature = gl_quad,
      inner_linear_method = "petsc_gmres",
      l_abs_tol = 1.0e-9,
      l_max_its = 300,
      gmres_restart_interval = 30,
    },
  },
  xs_map = xs_map,
  options = {
    scattering_order = 0,
    spatial_discretization = "pwld",
    boundary_conditions = { { name = "zmin", type = "vacuum" }, { name = "zmax", type = "vacuum" } },
    save_angular_flux = true,
    volumetric_sources = { src0, src1 },
  },
}

phys1 = lbs.DiscreteOrdinatesProblem.Create(lbs_block)

-- Initialize and execute solver
ss_solver = lbs.SteadyStateSolver.Create({ lbs_problem = phys1 })

ss_solver:Initialize()
ss_solver:Execute()

leakage_left_1 = lbs.ComputeLeakage(phys1, { "zmin" })["zmin"][1]
leakage_right_1 = lbs.ComputeLeakage(phys1, { "zmax" })["zmax"][1]

lbs.WriteAngularFluxes(phys1, "angular_io")

phys2 = lbs.DiscreteOrdinatesProblem.Create(lbs_block)

ss_solver_2 = lbs.SteadyStateSolver.Create({ lbs_problem = phys2 })

ss_solver_2:Initialize()

lbs.ReadAngularFluxes(phys2, "angular_io")

leakage_left_2 = lbs.ComputeLeakage(phys2, { "zmin" })["zmin"][1]
leakage_right_2 = lbs.ComputeLeakage(phys2, { "zmax" })["zmax"][1]

leakage_left_diff = leakage_left_1 - leakage_left_2
leakage_right_diff = leakage_right_1 - leakage_right_2

log.Log(LOG_0, string.format("Leakage-Diff1=%.5e", leakage_left_diff))
log.Log(LOG_0, string.format("Leakage-Diff2=%.5e", leakage_right_diff))

os.execute("rm angular_io0.h5")

File: test/python/modules/linear_boltzmann_solvers/transport_steady/angular_io_1d.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Standard Reed 1D 1-group problem

import os
import sys
import numpy as np

if "opensn_console" not in globals():
    from mpi4py import MPI
    size = MPI.COMM_WORLD.size
    rank = MPI.COMM_WORLD.rank
    sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../../")))
    from pyopensn.mesh import OrthogonalMeshGenerator
    from pyopensn.xs import MultiGroupXS
    from pyopensn.source import VolumetricSource
    from pyopensn.aquad import GLProductQuadrature1DSlab
    from pyopensn.solver import DiscreteOrdinatesProblem, SteadyStateSourceSolver
    from pyopensn.logvol import RPPLogicalVolume

if __name__ == "__main__":

    # Create Mesh
    widths = [2., 1., 2., 1., 2.]
    nrefs = [200, 200, 200, 200, 200]
    Nmat = len(widths)
    nodes = [0.]
    for imat in range(Nmat):
        dx = widths[imat] / nrefs[imat]
        for i in range(nrefs[imat]):
            nodes.append(nodes[-1] + dx)
    meshgen = OrthogonalMeshGenerator(node_sets=[nodes])
    grid = meshgen.Execute()

    # Set block IDs
    z_min = 0.0
    z_max = widths[1]
    for imat in range(Nmat):
        z_max = z_min + widths[imat]
        if rank == 0:
            print("imat=", imat, ", zmin=", z_min, ", zmax=", z_max)
        lv = RPPLogicalVolume(infx=True, infy=True, zmin=z_min, zmax=z_max)
        grid.SetBlockIDFromLogicalVolume(lv, imat, True)
        z_min = z_max

    # Add cross sections to materials
    total = [50., 5., 0., 1., 1.]
    c = [0., 0., 0., 0.9, 0.9]
    xs_map = len(total) * [None]
    for imat in range(Nmat):
        xs_ = MultiGroupXS()
        xs_.CreateSimpleOneGroup(total[imat], c[imat])
        xs_map[imat] = {
            "block_ids": [imat], "xs": xs_,
        }

    # Create sources in 1st and 4th materials
    src0 = VolumetricSource(block_ids=[0], group_strength=[50.])
    src1 = VolumetricSource(block_ids=[3], group_strength=[1.])

    # Angular Quadrature
    gl_quad = GLProductQuadrature1DSlab(n_polar=128, scattering_order=0)

    # LBS block option
    num_groups = 1
    solver_dict = {}
    solver_dict["mesh"] = grid
    solver_dict["num_groups"] = num_groups
    solver_dict["groupsets"] = [
        {
            "groups_from_to": (0, num_groups - 1),
            "angular_quadrature": gl_quad,
            "inner_linear_method": "petsc_gmres",
            "l_abs_tol": 1.0e-9,
            "l_max_its": 300,
            "gmres_restart_interval": 30,
        },
    ]
    solver_dict["xs_map"] = xs_map
    solver_dict["volumetric_sources"] = [src0, src1]
    solver_dict["boundary_conditions"] = [
        {"name": "zmin", "type": "vacuum"},
        {"name": "zmax", "type": "vacuum"}
    ]
    solver_dict["options"] = {
        "save_angular_flux": True,
    }

    # Initialize and execute solvers
    phys1 = DiscreteOrdinatesProblem(**solver_dict)
    ss_solver = SteadyStateSourceSolver(problem=phys1)
    ss_solver.Initialize()
    ss_solver.Execute()
    psi1 = phys1.GetPsi()
    phys1.WriteAngularFluxes("angular_io")

    phys2 = DiscreteOrdinatesProblem(**solver_dict)
    ss_solver_2 = SteadyStateSourceSolver(problem=phys2)
    ss_solver_2.Initialize()
    phys2.ReadAngularFluxes("angular_io")
    psi2 = phys2.GetPsi()

    for i, (arr1, arr2) in enumerate(zip(psi1, psi2)):
        if arr1.shape != arr2.shape:
            raise ValueError(f"Different shapes at index {i}: {arr1.shape} vs {arr2.shape}")

    if not np.allclose(arr1, arr2, rtol=1e-8, atol=1e-12):
        diff = np.max(np.abs(arr1 - arr2))
        print(f"Difference at index {i}, max diff = {diff}")
        raise ValueError(f"psi mismatch at index {i}")

    print(f"psi values match for rank {rank}")

    os.remove(f"angular_io{rank}.h5")
